File: packages/w2v/reference/src/training/objective.c
#include "training.h"

void objective_train(
    ModelStep *step,
    const real *hidden,
    real *hidden_gradient)
{
    if (step->trainer->config.objective_kind ==
        OBJECTIVE_HIERARCHICAL_SOFTMAX)
    {
        hierarchical_softmax_train(step, hidden, hidden_gradient);
        return;
    }

    negative_sampling_train(step, hidden, hidden_gradient);
}

File: packages/w2v/src/training/objective.rs
use super::{hierarchical_softmax, negative_sampling};
use crate::{
    config::{ObjectiveKind, Real},
    random::Rng,
    trainer::Trainer,
};

pub fn train(
    trainer: &Trainer,
    target_token: usize,
    learning_rate: Real,
    negative_rng: &mut Rng,
    hidden: &[Real],
    hidden_gradient: &mut [Real],
) {
    if trainer.config.objective_kind == ObjectiveKind::HierarchicalSoftmax {
        hierarchical_softmax::train(
            trainer,
            target_token,
            learning_rate,
            hidden,
            hidden_gradient,
        );
    } else {
        negative_sampling::train(
            trainer,
            target_token,
            learning_rate,
            negative_rng,
            hidden,
            hidden_gradient,
        );
    }
}
